darshan-util/pydarshan/darshan/backend/cffi_backend.py

```diff
@@ -337,49 +339,45 @@ def log_get_generic_record(log, mod_name, dtype='numpy'):
     if mod_name == 'H5D':
         rec['file_rec_id'] = rbuf[0].file_rec_id
 
-    clst = []
-    for i in range(0, len(rbuf[0].counters)):
-        clst.append(rbuf[0].counters[i])
-    rec['counters'] = np.array(clst, dtype=np.int64)
-    cdict = dict(zip(counter_names(mod_name), rec['counters']))
-
-    flst = []
-    for i in range(0, len(rbuf[0].fcounters)):
-        flst.append(rbuf[0].fcounters[i])
-    rec['fcounters'] = np.array(flst, dtype=np.float64)
-    fcdict = dict(zip(fcounter_names(mod_name), rec['fcounters']))
-
-    if dtype == "dict":
-        rec.update({
-            'counters': cdict, 
-            'fcounters': fcdict
-            })
-
-    if dtype == "pandas":
-        rec['id'] = np.uint64(rec['id'])
-        df_c = pd.DataFrame(cdict, index=[0])
-        df_fc = pd.DataFrame(fcdict, index=[0])
-
-        # flip column order (to prepend id and rank)
-        df_c = df_c[df_c.columns[::-1]]
-        df_fc = df_fc[df_fc.columns[::-1]]
-
-        # attach id and rank to counters and fcounters
-        df_c['id'] = rec['id']
-        df_c['rank'] = rec['rank']
-
-        df_fc['id'] = rec['id']
-        df_fc['rank'] = rec['rank']
-
-        # flip column order
-        df_c = df_c[df_c.columns[::-1]]
-        df_fc = df_fc[df_fc.columns[::-1]]
-
-        rec.update({
-            'counters': df_c,
-            'fcounters': df_fc
-            })
-
+    clst = np.zeros(len(rbuf[0].counters), dtype=np.int64)
+    for i in range(clst.size):
+        clst[i] = rbuf[0].counters[i]
+
+    flst = np.zeros(len(rbuf[0].fcounters), dtype=np.float64)
+    for i in range(flst.size):
+        flst[i] = rbuf[0].fcounters[i]
```

Comment:
These loops are not needed. I think I did mention in the past that the CFFI documentation is worth looking over a bit. Here's a first draft that passes tests and improves performance relative to this PR branch, but probably just scratches the surface of what may be possible by using low-level buffers/pointers:

`asv continuous -b "time_mod_read_all_records" nawtrey_issue_440_get_generic_record_perf treddy_cffi_buffers`

```
       before           after         ratio
     [f602e959]       [669085bc]
     <nawtrey_issue_440_get_generic_record_perf>       <treddy_cffi_buffers>
-     1.03±0.03ms         854±70μs     0.83  report.ModReadAllRecords.time_mod_read_all_records('dict', 'tests/input/sample-dxt-simple.darshan', 'MPI-IO', True)
-        898±70μs         709±20μs     0.79  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'tests/input/sample-dxt-simple.darshan', 'MPI-IO', True)
-        752±20μs         514±20μs     0.68  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'tests/input/sample-dxt-simple.darshan', 'MPI-IO', False)
-     2.20±0.05ms      1.42±0.08ms     0.64  report.ModReadAllRecords.time_mod_read_all_records('dict', 'tests/input/sample-dxt-simple.darshan', 'POSIX', True)
-     2.10±0.07ms       1.34±0.1ms     0.64  report.ModReadAllRecords.time_mod_read_all_records('dict', 'tests/input/sample-dxt-simple.darshan', 'POSIX', False)
-      1.79±0.1ms         945±20μs     0.53  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'tests/input/sample-dxt-simple.darshan', 'POSIX', True)
-     1.74±0.09ms          899±3μs     0.52  report.ModReadAllRecords.time_mod_read_all_records('numpy', 'tests/input/sample-dxt-simple.darshan', 'POSIX', False)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

```diff
diff --git a/darshan-util/pydarshan/darshan/backend/cffi_backend.py b/darshan-util/pydarshan/darshan/backend/cffi_backend.py
index 1ae6159..8291495 100644
--- a/darshan-util/pydarshan/darshan/backend/cffi_backend.py
+++ b/darshan-util/pydarshan/darshan/backend/cffi_backend.py
@@ -339,13 +339,8 @@ def log_get_generic_record(log, mod_name, c_cols=None, fc_cols=None, dtype='nump
     if mod_name == 'H5D':
         rec['file_rec_id'] = rbuf[0].file_rec_id
 
-    clst = np.zeros(len(rbuf[0].counters), dtype=np.int64)
-    for i in range(clst.size):
-        clst[i] = rbuf[0].counters[i]
-
-    flst = np.zeros(len(rbuf[0].fcounters), dtype=np.float64)
-    for i in range(flst.size):
-        flst[i] = rbuf[0].fcounters[i]
+    clst = np.frombuffer(ffi.buffer(rbuf[0].counters), dtype=np.int64)
+    flst = np.frombuffer(ffi.buffer(rbuf[0].fcounters), dtype=np.float64)
 
     if c_cols is None:
         c_cols = counter_names(mod_name)
```

Reply:
Thanks for the help here, this boosted performance and simplified the code a bit.  

I implemented these changes in the latest version, so I'll resolve this.